### ISSUE TITLE
Sprites & 2D rendering

### ISSUE THREAD

**Issue by BearishSun** [GameFoundry/bsf#37]
Add the ability to easily define 2D sprites from textures, and add support for pixel perfect rendering (requires enhancement of the scene object system). (Sprites are also required by the particle system, so it might be implemented there).
Tile map rendering (with import from Tiled app)
Layer rendering (for parallax)

**Comment by nemerle** [comment]
Would drawing in-scene text quads fall under this ?

**Comment by BearishSun** [comment]
I haven't fully thought out how would the 2D system work, but it would likely in some way work with the GUI system (quite possibly the GUI system would be re-worked to use the lower-level sprite rendering system, or the current GUI-specific sprite rendering would be extended to support general purpose 2D). Either way text rendering in 2D scenes would be supported.

This would not include text rendering in 3D scenes (e.g. text on the walls, etc.).

**Comment by nemerle** [comment]
Well, text-as-decals is one use-case, another is floating text above some other scene object ( damage notifications, object names etc. )

**Comment by BearishSun** [comment]
Indeed, although I suspect pretty much all of the code between those use cases will be shared. Technically you can do 3D text right now, by applying an arbitrary 3D transformation on `GUIWidget` but I do want to make it more straightforward (i.e. a `Text` component or similar). 

Also `TextSprite` along with `TextData` provide a way to convert text + font into a set of sprites, which can then be rendered using the low level rendering API. (This is what GUI uses internally)

Ideally we should also provide support for distance field text rendering, as it looks really nice with large text sizes (esp. for decals), doesn't require separate font bitmaps for different sizes and takes up much less space.
